polkadot/node/collation-generation/src/lib.rs

```diff
@@ -212,57 +212,90 @@ async fn handle_new_activations<Context>(
 		);
 
 		let availability_cores = availability_cores??;
-		let n_validators = validators??.len();
 		let async_backing_params = async_backing_params?.ok();
+		let n_validators = validators??.len();
+		let para_backing_state =
+			para_backing_state??.ok_or(crate::error::Error::MissingParaBackingState)?;
+
 		let maybe_claim_queue = fetch_claim_queue(ctx.sender(), relay_parent).await?;
 
-		for (core_idx, core) in availability_cores.into_iter().enumerate() {
-			let _availability_core_timer = metrics.time_new_activations_availability_core();
+		// The loop bellow will fill in cores that the para is allowed to build on.
+		let mut cores_to_build_on = Vec::new();
 
-			let (scheduled_core, assumption) = match core {
-				CoreState::Scheduled(scheduled_core) =>
-					(scheduled_core, OccupiedCoreAssumption::Free),
+		for (core_idx, core) in availability_cores.into_iter().enumerate() {
+			match core {
+				CoreState::Scheduled(scheduled_core) => {
+					// Core is scheduled and free, but is it ours ?
+					if scheduled_core.para_id != config.para_id {
+						gum::trace!(
+							target: LOG_TARGET,
+							core_idx = %core_idx,
+							relay_parent = ?relay_parent,
+							our_para = %config.para_id,
+							their_para = ?scheduled_core.para_id,
+							"core is not assigned to our para. Keep going.",
+						);
+						continue
+					}
+				},
 				CoreState::Occupied(occupied_core) => match async_backing_params {
 					Some(params) if params.max_candidate_depth >= 1 => {
 						// maximum candidate depth when building on top of a block
 						// pending availability is necessarily 1 - the depth of the
 						// pending block is 0 so the child has depth 1.
 
-						// TODO [now]: this assumes that next up == current.
-						// in practice we should only set `OccupiedCoreAssumption::Included`
-						// when the candidate occupying the core is also of the same para.
-						let res = match maybe_claim_queue {
-							Some(ref claim_queue) => {
-								// read what's in the claim queue for this core
-								fetch_next_scheduled_on_core(
-									claim_queue,
-									CoreIndex(core_idx as u32),
-								)
-							},
-							None => {
-								// Runtime doesn't support claim queue runtime api. Fallback to
-								// `next_up_on_available`
-								occupied_core.next_up_on_available
-							},
-						}
-						.map(|scheduled| (scheduled, OccupiedCoreAssumption::Included));
-
-						match res {
-							Some(res) => res,
-							None => continue,
+						// Use claim queue if available, or fallback to `next_up_on_available`
+						if let Some(ref claim_queue) = maybe_claim_queue {
+							// Check if our para is scheduled assuming the candidate pending
+							// availability gets included in the very next block
+							let next_para_scheduled = fetch_next_scheduled_on_core(
+								&claim_queue,
+								CoreIndex(core_idx as _),
+							);
+							if Some(config.para_id) != next_para_scheduled {
```

Comment:
Is it intended that if claim_queue is `Some`, we always `continue` here? There should be a branch that checks for para id equality and does not continue no?

Reply:
No, sorry, was temporary code, I've not merged it properly, and just accepted my changes over some other changes. Take a look now